schema-layer/data-structures/generics.md

```diff
@@ -0,0 +1,79 @@
+# Specification: IPLD Generics
```

Comment:
Once I started writing this up the name “multi-block types” didn’t make a lot of sense. A fair amount of what we’re doing is actually contained in a single block. Also, once I started to back out the `kind` inheritance hierarchy it was much more obviously a system for generics so I went with this naming, but the name is still up for debate so if people hate it let me know.

Reply:
We can't call these generics. A "generic" type is a type constructor or a type parametrized by another type. Calling these "generics" will confuse everyone.

Alternatives:

* Composite types
* Complex types
* Collection
* Composite Kinds

None of these are perfect but Generics is a no go (like, we-need-to-erase-every-mention-of-it no-go or we'll be answering questions on this till the end of time).

Reply:
I like “composite” because I don’t think I’ve ever seen it used as a name for anything before, so I’m +1 on composite types.

Reply:
"compound" is in a similar area.

Let's not totally throw out "advanced layouts" either, I think it has merit here.

Reply:
Most of the spec/system is a programmatic interface, so it seem odds to refer them as “layouts.”

Reply:
I still like "advanced layout". To me that means data structures laid out in an advanced way. Think e.g. also about the use case were you want to store a big object and just want to split it internally into a maximum sized nodes (to get your stuff working with Bitswap).

Reply:
yeah, "advanced layout" is "this block isn't what it looks like"

<img width="257" alt="Screen Shot 2019-06-14 at 8 54 25 pm" src="https://user-images.githubusercontent.com/495647/59504609-abf39380-8ee6-11e9-9ebb-c124e8033082.png">

Reply:
> yeah, "advanced layout" is "this block isn't what it looks like"

Sure, but users of this system are moving past thinking in a block specific way, so it feels rather odd to describe that system in terms of the block layout since that’s exactly the mental model we’re trying to push them out of.

Reply:
It's not necessarily about blocks. I think the user of IPLD still needs to be aware that their data is split into chunks (whether we call them "blocks" or not). So it's rather a "this piece of data isn't what it looks like, but has a different shape than you might think".

Reply:
I'm warmly neutral to "composite" and "complex" as part of the wording we settle on (which is about how I'd describe my overall warmth to 'advanced layouts' as well).

Bit dislike putting "types" or "kinds" immediately after the other word though, as it is in several of steb's examples above.

e.g. "composite kinds" is a bit odd -- it's not the kinds that are composite; the overall thing still acts like it a regular unsurprising not-novel kind from the enum of kinds we already know and love.

Moar brainstorming ftw?  I've got a piece of paper on which I also jotted "prisms" for example.  (Terrible?  Maybe.  Brainstorming?  Does go through phases of such...)

Reply:
Yeah, "prisms" is like "lens" which is a concept I like here, "the lens through which you can turn these raw crazy blocks into useful, traveresable, queryable kinds". Probably a bit too esoteric for naming purposes though.

Reply:
Remember that we justified the use of "kinds" with the idea that "types" in the schema layer was where most of the interesting action was going to be. Our current trajectory has us splitting off from schemas into this new place, entirely detached. So we no longer have a clear hierarchy and may end up having schema "types" being a more obscure thing that's only used when we invoke schemas for specifying a certain subsets of data layouts. We're currently heading to a place where we will mostly be talking about data layer kinds and these advanced layout / generic things most of the time, and "types" only occasionally.
We need to find the synthesis with schemas here. And if not, maybe "types" is back on the table and schemas can get something less general because they're in the corner doing a specific, limited job.

Reply:
> Yeah, "prisms" is like "lens" which is a concept I like here

Before someone starts to like the idea of calling it "lens", I'm strongly oppose it. It might be niche, but it's a term from bi-directional transformations. I even wrote a JS lib implementing those: https://github.com/vmx/jslens

Reply:
renamed to composites.